### ISSUE TITLE
`clippy::exit` should ignore foreign-macro expansions

### ISSUE THREAD

**Issue by sandersaares** [rust-lang/rust-clippy#17082]
### Summary

`clippy::exit` fires on `std::process::exit(...)` calls that come from
macro expansions, even when the macro lives in a different crate that the
user has no control over. This makes it impossible for a downstream user
to enable `#![deny(clippy::exit)]` (or even leave it as the default
restriction-group warning) in a crate that uses such a macro.

A user who wants to enforce "no exits in my code" cannot do so without
suppressing the lint at the macro call site, which also suppresses checks
for any user-written `exit(...)` calls in the same scope.

I suggest that `clippy::exit` should treat `exit(...)` calls that come
from foreign-crate macro expansions as out of scope by default — `clippy`
already does this for several other restriction lints by gating them on
`in_external_macro` / `is_from_proc_macro`. The user can fix their own
code; they cannot fix code expanded by a third-party macro.

### Lint Name

exit

### I tried this code:

```rust
#![deny(clippy::exit)]

// Imagine this macro lives in a third-party crate you do not control.
macro_rules! exit_via_macro {
    () => {
        std::process::exit(0);
    };
}

fn cleanup_and_quit() {
    exit_via_macro!();
}

fn main() {
    cleanup_and_quit();
}
```

(One-click playground:
<https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=ade8d098751b0b86df1c91abaf0511a1>
— then **Tools → Clippy**.)

### I saw this happen:

```
error: usage of `process::exit`
  --> src/main.rs:5:9
   |
 5 |         std::process::exit(0);
   |         ^^^^^^^^^^^^^^^^^^^^^
...
10 |     exit_via_macro!();
   |     ----------------- in this macro invocation
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#exit
note: the lint level is defined here
  --> src/main.rs:1:9
   |
 1 | #![deny(clippy::exit)]
   |         ^^^^^^^^^^^^
   = note: this error originates in the macro `exit_via_macro` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not compile `single` (bin "single") due to 1 previous error
```

### I expected to see this happen:

No diagnostic. The `exit(...)` call originates from inside a foreign
`macro_rules!` body that the downstream user cannot edit, so the lint
has no actionable target. As an alternative, a `clippy.toml` allowlist
for macro paths whose expansions should be exempt would also be
acceptable, but exempting foreign-macro expansions by default would
match what several other restriction lints already do via
`in_external_macro` / `is_from_proc_macro`.

The only currently available workarounds are `#[allow(clippy::exit)]` or
`#[expect(clippy::exit)]` at the macro call site, both of which also
suppress any user-written `exit(...)` calls in the same scope.

### Version

```text
- Reproducible on rustc 1.95.0 stable (clippy 0.1.95) — both
  `x86_64-unknown-linux-gnu` and `x86_64-pc-windows-msvc`.
- Also reproducible on rustc 1.93.0 stable (clippy 0.1.93) on Linux.
- Also reproducible on nightly 1.98.0-nightly (2026-05-27, clippy 0.1.97)
  on both Linux and Windows — i.e. the lint design has not changed on
  master.
- Rust 2024 edition.

rustc 1.93.1 (01f6ddf75 2026-02-11)
binary: rustc
commit-hash: 01f6ddf7588f42ae2d7eb0a2f21d44e8e96674cf
commit-date: 2026-02-11
host: x86_64-pc-windows-msvc
release: 1.93.1
LLVM version: 21.1.8
```

**Comment by y21** [comment]
Do you have a practical example of a crate with a macro that expands to an exit() call and that you wouldn't want it to lint? I'm having a hard time convincing myself that suppressing is the right solution (but it may well be -- personally I don't use this lint which makes it a bit harder to know what makes sense).

In the given reproducer, it seems basically equivalent to calling `exit()` and I don't know why you wouldn't want it to lint.
From reading the lint description, it seems to me as though the motivation for enabling it is to find calls to `exit()` outside of main as by itself it gives no hints why the program exited, so the lint in essence is correct and the motivation still applies to the reproducer.
(In a similar way, correctness lints also often want to lint external code as reporting bugs is useful even if it's not actionable at all)

Unless the main argument here is that you can't `#[allow]` it, in which case, can you elaborate on that? I understand that `#[allow]` on the macro invocation directly doesn't work as it doesn't propagate to the expansion, so are you trying to put it on the enclosing block/fn? I guess then yes that may allow other code, but you could also put it into its own block alone (or use some other syntax).
```rs
#[allow(clippy::exit)]
{ exit_via_macro!(); }
```
I do agree though, it's not great, but this does still let you allow it individually.

**Comment by samueltardieu** [comment]
My reasoning when opening #17105 was that we wouldn't lint an external wrapper function that would just forward its argument to `exit()`, as we don't do whole program analysis. The external macro seems to be a slight variation of such a situation.

**Comment by sandersaares** [comment]
This is what motivated the issue report: https://github.com/gungraun/gungraun/blob/78a622af693ca15f8221f24c72f60958370f58e0/gungraun/src/macros.rs#L389

**Comment by y21** [comment]
Considering that this macro is intended to generate a sort of "main" fn I wonder if the lint should be `#[allow]`ed in there since the lint wants to catch non-obvious exit calls outside of main. But yeah, I guess that is a better example. I can see the motivation and how there might be more crates that do something like that. Not all crate authors want to litter their macros with a bunch of `#[allow]`s to satisfy restriction lints in downstream crates
